File: timer_manager.cs
// timer_manager.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Threading;
using System.Threading.Tasks;
using Newtonsoft.Json;

class Timer
{
    public string Name { get; set; }
    public double Duration { get; set; }
    public double Elapsed { get; set; }
    public bool Running { get; set; }
    public bool Finished { get; set; }

    public double Remaining() => Finished ? 0 : Math.Max(0, Duration - Elapsed);

    public bool Update(double dt)
    {
        if (Running && !Finished)
        {
            Elapsed += dt;
            if (Elapsed >= Duration)
            {
                Elapsed = Duration;
                Running = false;
                Finished = true;
                return true;
            }
        }
        return false;
    }
}

class TimerManager
{
    private List<Timer> timers = new List<Timer>();
    private int selected = 0;
    private bool running = true;
    private string saveFile = "timers.json";
    private DateTime lastUpdate = DateTime.Now;

    public TimerManager()
    {
        Load();
        if (timers.Count == 0)
        {
            timers.Add(new Timer { Name = "Pomodoro", Duration = 25 * 60 });
            timers.Add(new Timer { Name = "Break", Duration = 5 * 60 });
        }
    }

    public void AddTimer(string name, double duration)
    {
        timers.Add(new Timer { Name = name, Duration = duration });
        Save();
    }

    public void RemoveTimer(int idx)
    {
        if (idx < 0 || idx >= timers.Count) return;
        timers.RemoveAt(idx);
        if (selected >= timers.Count) selected = timers.Count - 1;
        Save();
    }

    public void Toggle(int idx)
    {
        if (idx < 0 || idx >= timers.Count) return;
        var t = timers[idx];
        if (t.Finished)
        {
            t.Elapsed = 0;
            t.Finished = false;
            t.Running = true;
        }
        else
        {
            t.Running = !t.Running;
        }
        Save();
    }

    public void Reset(int idx)
    {
        if (idx < 0 || idx >= timers.Count) return;
        var t = timers[idx];
        t.Elapsed = 0;
        t.Running = false;
        t.Finished = false;
        Save();
    }

    public void Update()
    {
        var now = DateTime.Now;
        double dt = (now - lastUpdate).TotalSeconds;
        lastUpdate = now;
        foreach (var t in timers)
        {
            if (t.Update(dt))
            {
                Console.WriteLine($"\n!!! Timer '{t.Name}' finished !!!");
            }
        }
    }

    public void Save()
    {
        string json = JsonConvert.SerializeObject(timers, Formatting.Indented);
        File.WriteAllText(saveFile, json);
    }

    public void Load()
    {
        if (File.Exists(saveFile))
        {
            string json = File.ReadAllText(saveFile);
            timers = JsonConvert.DeserializeObject<List<Timer>>(json) ?? new List<Timer>();
        }
    }

    public void Run()
    {
        Console.CursorVisible = false;
        Console.Clear();
        lastUpdate = DateTime.Now;

        while (running)
        {
            Console.SetCursorPosition(0, 0);
            Console.WriteLine("=== Timer Manager ===");
            Console.WriteLine($"Timers: {timers.Count}   [a]dd  [d]elete  [Space]toggle  [r]eset  [q]uit");

            for (int i=0; i<timers.Count; i++)
            {
                var t = timers[i];
                double rem = t.Remaining();
                string status = t.Finished ? " ✓" : (t.Running ? " ▶" : " ⏸");
                string line = $"{i+1}. {t.Name} {status}  {rem/60:00}:{rem%60:00} / {t.Duration/60:00}:{t.Duration%60:00}";
                if (i == selected)
                {
                    Console.BackgroundColor = ConsoleColor.White;
                    Console.ForegroundColor = ConsoleColor.Black;
                }
                else
                {
                    Console.ResetColor();
                    if (t.Running) Console.ForegroundColor = ConsoleColor.Green;
                    else if (t.Finished) Console.ForegroundColor = ConsoleColor.Red;
                    else Console.ForegroundColor = ConsoleColor.Yellow;
                }
                Console.WriteLine(line);
                Console.ResetColor();
            }
            Console.WriteLine("\nCommands: a=add, d=delete, Space=toggle, r=reset, q=quit");

            if (Console.KeyAvailable)
            {
                var key = Console.ReadKey(true);
                switch (key.Key)
                {
                    case ConsoleKey.Q: running = false; break;
                    case ConsoleKey.A:
                        Console.Write("Enter name: ");
                        string name = Console.ReadLine();
                        Console.Write("Enter duration (seconds): ");
                        if (double.TryParse(Console.ReadLine(), out double dur))
                            AddTimer(name, dur);
                        break;
                    case ConsoleKey.D:
                        if (timers.Count > 0) RemoveTimer(selected);
                        break;
                    case ConsoleKey.Spacebar:
                        if (timers.Count > 0) Toggle(selected);
                        break;
                    case ConsoleKey.R:
                        if (timers.Count > 0) Reset(selected);
                        break;
                    case ConsoleKey.UpArrow:
                        if (selected > 0) selected--;
                        break;
                    case ConsoleKey.DownArrow:
                        if (selected < timers.Count-1) selected++;
                        break;
                }
            }

            Update();
            Thread.Sleep(50);
        }
        Save();
        Console.CursorVisible = true;
    }

    static void Main()
    {
        var tm = new TimerManager();
        tm.Run();
    }
}

File: timer_manager.rs
// timer_manager.rs
use crossterm::{
    cursor, event, terminal, ExecutableCommand, QueueableCommand, Result,
    style::{Color, Print, ResetColor, SetForegroundColor, SetBackgroundColor},
};
use serde::{Deserialize, Serialize};
use std::fs;
use std::io::{stdout, Write};
use std::time::{Duration, Instant};

#[derive(Serialize, Deserialize, Clone)]
struct Timer {
    name: String,
    duration: f64, // seconds
    elapsed: f64,
    running: bool,
    finished: bool,
}

impl Timer {
    fn remaining(&self) -> f64 {
        if self.finished { 0.0 } else { (self.duration - self.elapsed).max(0.0) }
    }
    fn update(&mut self, dt: f64) -> bool {
        if self.running && !self.finished {
            self.elapsed += dt;
            if self.elapsed >= self.duration {
                self.elapsed = self.duration;
                self.running = false;
                self.finished = true;
                return true;
            }
        }
        false
    }
}

struct TimerManager {
    timers: Vec<Timer>,
    selected: usize,
    save_file: String,
    running: bool,
}

impl TimerManager {
    fn new(save_file: &str) -> Self {
        let mut tm = TimerManager {
            timers: Vec::new(),
            selected: 0,
            save_file: save_file.to_string(),
            running: true,
        };
        tm.load();
        if tm.timers.is_empty() {
            tm.timers.push(Timer { name: "Pomodoro".to_string(), duration: 25.0*60.0, elapsed: 0.0, running: false, finished: false });
            tm.timers.push(Timer { name: "Break".to_string(), duration: 5.0*60.0, elapsed: 0.0, running: false, finished: false });
        }
        tm
    }

    fn add_timer(&mut self, name: String, duration: f64) {
        self.timers.push(Timer { name, duration, elapsed: 0.0, running: false, finished: false });
        self.save();
    }

    fn remove_timer(&mut self, idx: usize) {
        if idx < self.timers.len() {
            self.timers.remove(idx);
            if self.selected >= self.timers.len() { self.selected = self.timers.len().saturating_sub(1); }
            self.save();
        }
    }

    fn toggle(&mut self, idx: usize) {
        if idx < self.timers.len() {
            let t = &mut self.timers[idx];
            if t.finished {
                t.elapsed = 0.0;
                t.finished = false;
                t.running = true;
            } else {
                t.running = !t.running;
            }
            self.save();
        }
    }

    fn reset(&mut self, idx: usize) {
        if idx < self.timers.len() {
            let t = &mut self.timers[idx];
            t.elapsed = 0.0;
            t.running = false;
            t.finished = false;
            self.save();
        }
    }

    fn update(&mut self, dt: f64) {
        for t in &mut self.timers {
            if t.update(dt) {
                // notification
                println!("\n!!! Timer '{}' finished !!!", t.name);
            }
        }
    }

    fn save(&self) {
        let data = serde_json::to_string_pretty(&self.timers).unwrap();
        fs::write(&self.save_file, data).ok();
    }

    fn load(&mut self) {
        if let Ok(data) = fs::read_to_string(&self.save_file) {
            if let Ok(timers) = serde_json::from_str(&data) {
                self.timers = timers;
            }
        }
    }

    fn run(&mut self) -> Result<()> {
        let mut stdout = stdout();
        terminal::enable_raw_mode()?;
        stdout.execute(cursor::Hide)?;

        let mut last_update = Instant::now();

        while self.running {
            stdout.execute(terminal::Clear(terminal::ClearType::All))?;
            stdout.queue(cursor::MoveTo(0, 0))?.queue(Print("=== Timer Manager ==="))?;
            stdout.queue(cursor::MoveTo(0, 1))?.queue(Print(format!("Timers: {}   [a]dd  [d]elete  [Space]toggle  [r]eset  [q]uit", self.timers.len())))?;

            let mut line = 3;
            for (i, t) in self.timers.iter().enumerate() {
                let rem = t.remaining();
                let status = if t.finished { " ✓" } else if t.running { " ▶" } else { " ⏸" };
                let display = format!("{}. {} {}  {:02.0}:{:02.0} / {:02.0}:{:02.0}",
                    i+1, t.name, status,
                    rem/60.0, rem%60.0,
                    t.duration/60.0, t.duration%60.0);
                stdout.queue(cursor::MoveTo(0, line))?;
                if i == self.selected {
                    stdout.queue(SetBackgroundColor(Color::White))?.queue(SetForegroundColor(Color::Black))?
                          .queue(Print(display))?.queue(ResetColor)?;
                } else {
                    let color = if t.running { Color::Green } else if t.finished { Color::Red } else { Color::Yellow };
                    stdout.queue(SetForegroundColor(color))?.queue(Print(display))?.queue(ResetColor)?;
                }
                line += 1;
            }
            stdout.queue(cursor::MoveTo(0, line+1))?.queue(Print("Commands: a=add, d=delete, Space=toggle, r=reset, q=quit"))?;
            stdout.flush()?;

            // Non-blocking input
            if event::poll(Duration::from_millis(50))? {
                if let event::Event::Key(key) = event::read()? {
                    match key.code {
                        event::KeyCode::Char('q') => self.running = false,
                        event::KeyCode::Char('a') => {
                            // Упрощённо: ввод имени и длительности
                            // В реальном коде нужно организовать диалог, здесь пропустим для краткости
                        }
                        event::KeyCode::Char('d') => if !self.timers.is_empty() { self.remove_timer(self.selected); }
                        event::KeyCode::Char(' ') => if !self.timers.is_empty() { self.toggle(self.selected); }
                        event::KeyCode::Char('r') => if !self.timers.is_empty() { self.reset(self.selected); }
                        event::KeyCode::Up => if self.selected > 0 { self.selected -= 1; }
                        event::KeyCode::Down => if self.selected < self.timers.len()-1 { self.selected += 1; }
                        _ => {}
                    }
                }
            }

            // Update timers
            let now = Instant::now();
            let dt = now.duration_since(last_update).as_secs_f64();
            last_update = now;
            self.update(dt);
            std::thread::sleep(Duration::from_millis(50));
        }

        terminal::disable_raw_mode()?;
        stdout.execute(cursor::Show)?;
        Ok(())
    }
}

fn main() -> Result<()> {
    let mut manager = TimerManager::new("timers.json");
    manager.run()
}
